### ISSUE TITLE
enhance normandy api to support rel-eng and other dashboarding.

### ISSUE THREAD

**Issue by gregglind** [mozilla/normandy#950]
Spitballing:
- "proposed study" (vs. actual)
- "proposed end date ts" (rel-eng wants to know when to double check!)
- "disable ts" (using last rev ts instead)
- "previous iteration for all studies" (for allowing Most Recent Change for all recipes at once)
- promote "sampling pct" to a first class thing (currently parsing jexl)
- responsible party / owner for a study (using creator as a proxy currently)

```
8261 glind ~/gits/x-shield-inflight-dashboard [git:master?+]$ node  cli.js current.recipes.json
Enabled   Approved  Preference                                          Bug         Channels  Sample %  Last Rev       Proposed End   Owner
*         *         layout.css.servo.enabled                            1381147     ??              20  07/26/2017     not set        mgr
*         *         layout.frame_rate                                   1370253     ??               1  07/19/2017     not set        mgr
*         *         extensions.screenshots.system-disabled              1369150     ??             100  07/13/2017     not set        mgr
                    network.http.rcwn.enabled                           1381816     ??             100  07/26/2017     not set        mgr
                    dom.script_loader.bytecode_cache.enabled            1373241     ??              30  07/06/2017     not set        mgr
                    network.http.rcwn.enabled                           1367951     ??             100  06/27/2017     not set        unk
                    plugins.remember_infobar_dismissal                  1368141     ??             100  06/01/2017     not set        jga
                    extensions.shield-recipe-client.experiments.lazy_c  1365680     ??        unknown   05/23/2017     not set        mke
                    test.normandy.prefflip.defaultpref                  https://gi  ??               1  05/05/2017     not set        gli
```

**Comment by mythmon** [comment]
For non-functional metadata, such as proposed end date, and responsible part, I propose a "free form" YAML field (or other structured data) that you can put any metadata you want in. How does that sound?

As for disable ts, we already plan on changing the enabled flag to be per-revision, so this will be apparent in the revision history.

I'm not sure what you mean by "previous iteration for all studies". Can you elaborate on this?

Finally, making sampling percentage first class: I think I would rather go the opposite direction with this, and provide better tools for manipulating and processing filters. This is longer term though, and may be unreasonable.

**Comment by Osmose** [comment]
> Finally, making sampling percentage first class: I think I would rather go the opposite direction with this, and provide better tools for manipulating and processing filters. This is longer term though, and may be unreasonable.

+1 to this if we can afford waiting for it. For example, we've spitballed eventually storing filter expressions as objects instead of strings, that look something like:

```json
{
  "bucketSample": {start: 0, count: 20, total: 1000},
  "country": {code: "us"},
  "locale": {code: "en-US"},
  "jexl": {expression: "!normandy.isFirstRun"}
}
```

(Don't get caught up in this specific syntax, it's just an example. Imagine each of the fields having a dedicated UI instead of all being in one big JEXL string.)

Something like that allows us to build better editors for filter expressions and retain the flexibility of JEXL, while also being much more consumable by tooling.

**Comment by mythmon** [comment]
Revisiting this, here's the current state of things:

> proposed study" (vs. actual)

Experimenter handles the case of a proposed study, and Normandy handles actual studies

> proposed end date ts (rel-eng wants to know when to double check!)

We don't have anything specific for this, but I expect #1082 (general purpose notes) could be used for this.

>disable ts (using last rev ts instead)

As of #1183, we are now tracking all enable and disables. We'll expose this info in the API soon :tm: (#1363).

> previous iteration for all studies (for allowing Most Recent Change for all recipes at once)

I'm not sure what this means, but it sounds like it is mostly a view into the current state of things (as opposed to new data on the server). This would be well suited for Delivery Console or other external tools, but isn't a good fit for this repo anymore.

> promote "sampling pct" to a first class thing (currently parsing jexl)

Filter objects handles this well, as mentioned in the last two comments

> responsible party / owner for a study (using creator as a proxy currently)

Responsible Party / owner sounds like people that would often not have Normandy user accounts. If that's true, I think that notes would cover them well, instead of a structured field. We can explore this more in the future though.

---

In short, I think we are making slow progress towards this goal, and all of the work desired is covered in other issues or other tools/repositories. Given all that, I'm going to close this issue.
